### PR TITLE
Correct min_Δz for VerticallyStretchedGrids

### PR DESCRIPTION
**Pull request by tomchor**
Using `min_Δz(grid::VerticallyStretchedRectilinearGrid)` on GPU fails. The reason is that

```julia
julia> minimum(view(grid.Δzᵃᵃᶜ, 1:grid.Nz))
ERROR: scalar getindex is disallowed
```
fails on the GPU. The line should be `minimum(view(parent(grid.Δzᵃᵃᶜ), 1:grid.Nz))`.

This error had been there all along but I guess no one caught it. This commit should fix it, but should we also include a test that would catch it?

### CODE REVIEW

**Review by francispoulin — APPROVED** [review]
Good catch! 

Did you also want to add in a test for this in this PR?

**Comment by tomchor** [comment]
I'm not sure. I'm not very familiar with the testing part of Oceananigans, so I'm not sure what would be a good test for this. Just testing that `min_Δz` works seems too specific, on the other hand, creating a whole simulation with a stretched grid and running it with an LES and adaptive time-stepping just to test that seems wasteful.

I'm happy to just merge this for now and create a test for this in the future. Thoughts?

CC @glwagner @ali-ramadhan

**Comment by francispoulin** [comment]
I presume the time step wizard would use this new function?  If yes, then maybe we could add a test that uses the wizard for one time step.

**Comment by francispoulin** [comment]
'test_simulations.jl` has a function called `run_time_step_wizard_tests` that has several tests for a `RegularRectilinearGrid`.  We could add in a test tha tuses a stretched vertical grid.  Just one option that comes to mind.

**Comment by tomchor** [comment]
> `test_simulations.jl`has a function called`run_time_step_wizard_tests`that has several tests for a`RegularRectilinearGrid`. We could add in a test tha tuses a stretched vertical grid. Just one option that comes to mind.

Thanks for the suggestion. I added an extra test there with VerticallyStretchedGrids which would have failed before this bug fix. Let's see if everything passes

**Comment by glwagner** [comment]
> The line should be `minimum(view(parent(grid.Δzᵃᵃᶜ), 1:grid.Nz))`

This is not quite right. We want a view of the interior indices; the indices `1:grid.Nz` of the parent array covers some of the halo region at the bottom, and does not cover all the points near the top / surface.

It should be sufficient to compute the minimum of `parent(grid.Δzᵃᵃᶜ)` because the spacing in the halo region is extended from the interior points (and thus contains no new extrema).

**Comment by tomchor** [comment]
Hmm, I get zero spacing in the halo regions in my examples, which is why I used the view:

```julia
julia> using Oceananigans

julia> Nx  = Ny = Nz = 8
8

julia> S = 0.8 # Stretching factor
0.8

julia> hyperbolically_spaced_nodes(k) = -Lz-Lz*(tanh(S * ( (-(k-34) - 1) / Nz - 1)) / tanh(S))
hyperbolically_spaced_nodes (generic function with 1 method)

julia> grid = VerticallyStretchedRectilinearGrid(size = (Nx, Ny, Nz),
                                                 architecture = CPU(),
                                                 x = (0,64),
                                                 y = (0,64),
                                                 halo = (3, 3, 3),
                                                 z_faces = hyperbolically_spaced_nodes)
VerticallyStretchedRectilinearGrid{Float64, Periodic, Periodic, Bounded}
                   domain: x ∈ [0.0, 64.0], y ∈ [0.0, 64.0], z ∈ [-79.40339218489486, -76.41530207598367]
                 topology: (Periodic, Periodic, Bounded)
  resolution (Nx, Ny, Nz): (8, 8, 8)
   halo size (Hx, Hy, Hz): (3, 3, 3)
grid spacing (Δx, Δy, Δz): (8.0, 8.0, [min=0.17244642285916711, max=0.6621568834429468])

julia> grid.Δzᵃᵃᶜ
14-element OffsetArray(::Vector{Float64}, -2:11) with eltype Float64 with indices -2:11:
 0.17244642285916711
 0.17244642285916711
 0.17244642285916711
 0.17244642285916711
 0.20978434591127382
 0.2549831529775588
 0.3095903998229801
 0.37540677489580787
 0.45450231451088996
 0.5492198144905558
 0.6621568834429468
 0.17244642285916711
 0.0
 0.0

julia> minimum(parent(grid.Δzᵃᵃᶜ))
0.0

julia> minimum(view(parent(grid.Δzᵃᵃᶜ), 1:grid.Nz))
0.17244642285916711
```

**Comment by glwagner** [comment]
That's a bug. It's also not the only one:

https://github.com/CliMA/Oceananigans.jl/issues/1753

**Comment by tomchor** [comment]
Alright. Should I merge this as is for now and change later (I don't see much downsides in having an extra `view`), or should we wait for that bug to be fixed upstream?

**Comment by glwagner** [comment]
Let's do this right. The range 1:Nz is arbitrary (because halos can be arbitrarily sized) and does not include the surface points, which often contain the smallest spacing and thus the strictest CFL limitation.

#1754 will hopefully be merged quickly. Let's fix that, merge it into this PR, and then calculate a minimum over all cell spacings points in this PR so that it doesn't cause pain for anyone else.

**Comment by tomchor** [comment]
> #1754 will hopefully be merged quickly. Let's fix that, merge it into this PR, and then calculate a minimum over all cell spacings points in this PR so that it doesn't cause pain for anyone else.

Agreed. I didn't know you were gonna find and fix the bug so fast!

**Comment by glwagner** [comment]
It's all simple stuff... someone just has to sit down and do it...

**Comment by glwagner** [comment]
#1753 is a more serious issue because it means the grid geometry isn't quite right. I'm not 100% sure what the implications are but certainly it lowers the order of accuracy of calculations at the least. High order stuff on stretched grids is already somewhat suspect before we resolve #1704 though.

**Comment by tomchor** [comment]
I may be doing something wrong, but I still get zero spacing in halo regions even after #1754 was merged. The output changed a bit though (specifically the third-to-last term in `grid.Δzᵃᵃᶜ`.

```julia
julia> using Oceananigans

julia> Nx  = Ny = Nz = 8
8

julia> Lz = 32
32

julia> S = 0.8 # Stretching factor
0.8

julia> hyperbolically_spaced_nodes(k) = -Lz-Lz*(tanh(S * ( (-(k-34) - 1) / Nz - 1)) / tanh(S))
hyperbolically_spaced_nodes (generic function with 1 method)

julia> grid = VerticallyStretchedRectilinearGrid(size = (Nx, Ny, Nz),
                                                 architecture = CPU(),
                                                 x = (0,64),
                                                 y = (0,64),
                                                 halo = (3, 3, 3),
                                                 z_faces = hyperbolically_spaced_nodes)
VerticallyStretchedRectilinearGrid{Float64, Periodic, Periodic, Bounded}
                   domain: x ∈ [0.0, 64.0], y ∈ [0.0, 64.0], z ∈ [-79.40339218489486, -76.41530207598367]
                 topology: (Periodic, Periodic, Bounded)
  resolution (Nx, Ny, Nz): (8, 8, 8)
   halo size (Hx, Hy, Hz): (3, 3, 3)
grid spacing (Δx, Δy, Δz): (8.0, 8.0, [min=0.17244642285916711, max=0.6621568834429468])

julia> grid.Δzᵃᵃᶜ
14-element OffsetArray(::Vector{Float64}, -2:11) with eltype Float64 with indices -2:11:
 0.17244642285916711
 0.17244642285916711
 0.17244642285916711
 0.17244642285916711
 0.20978434591127382
 0.2549831529775588
 0.3095903998229801
 0.37540677489580787
 0.45450231451088996
 0.5492198144905558
 0.6621568834429468
 0.6621568834429468
 0.0
 0.0
```

@glwagner  I'm assuming the last four terms should all be  0.6621568834429468?

**Comment by tomchor** [comment]
CC @ali-ramadhan 

Fixed a bug in the construction of the halo regions that caused zero spacing in the top halo. Also changed `min_Δz` according to @glwagner's suggestion.

Here are the outputs for the grid now:

```julia
julia> using Oceananigans

julia> Nx  = Ny = Nz = 8
8

julia> Lz = 32
32

julia> S = 0.8 # Stretching factor
0.8

julia> hyperbolically_spaced_nodes(k) = -Lz-Lz*(tanh(S * ( (-(k-34) - 1) / Nz - 1)) / tanh(S))
hyperbolically_spaced_nodes (generic function with 1 method)

julia> grid = VerticallyStretchedRectilinearGrid(size = (Nx, Ny, Nz),
                                                 architecture = CPU(),
                                                 x = (0,64),
                                                 y = (0,64),
                                                 halo = (3, 3, 3),
                                                 z_faces = hyperbolically_spaced_nodes)
VerticallyStretchedRectilinearGrid{Float64, Periodic, Periodic, Bounded}
                   domain: x ∈ [0.0, 64.0], y ∈ [0.0, 64.0], z ∈ [-79.40339218489486, -76.41530207598367]
                 topology: (Periodic, Periodic, Bounded)
  resolution (Nx, Ny, Nz): (8, 8, 8)
   halo size (Hx, Hy, Hz): (3, 3, 3)
grid spacing (Δx, Δy, Δz): (8.0, 8.0, [min=0.17244642285916711, max=0.6621568834429468])

julia> grid.Δzᵃᵃᶜ
14-element OffsetArray(::Vector{Float64}, -2:11) with eltype Float64 with indices -2:11:
 0.17244642285916711
 0.17244642285916711
 0.17244642285916711
 0.17244642285916711
 0.20978434591127382
 0.2549831529775588
 0.3095903998229801
 0.37540677489580787
 0.45450231451088996
 0.5492198144905558
 0.6621568834429468
 0.6621568834429468
 0.6621568834429468
 0.6621568834429468

julia> grid.Δzᵃᵃᶠ
15-element OffsetArray(::Vector{Float64}, -3:11) with eltype Float64 with indices -3:11:
 0.17244642285916711
 0.17244642285916711
 0.17244642285916711
 0.17244642285916711
 0.19111538438522757
 0.2323837494444092
 0.28228677640026945
 0.3424985873593869
 0.4149545447033631
 0.5018610645007158
 0.6056883489667513
 0.6621568834429468
 0.6621568834429468
 0.6621568834429468
 0.6621568834429468
```

File: src/Grids/vertically_stretched_rectilinear_grid.jl [hunk 1]
```diff
@@ -206,7 +206,7 @@ function generate_stretched_vertical_grid(FT, z_topo, Nz, Hz, z_faces)
     z¹, zᴺ⁺¹ = interior_zF[1], interior_zF[Nz+1]
 
     zF₋ = [z¹   - sum(ΔzF₋[k:Hz]) for k = 1:Hz] # locations of faces in lower halo
-    zF₊ = [zᴺ⁺¹ + ΔzF₊[k]         for k = 1:Hz] # locations of faces in width of top halo region
+    zF₊ = reverse([zᴺ⁺¹ + sum(ΔzF₊[k:Hz]) for k = 1:Hz]) # locations of faces in width of top halo region
```
Comment (glwagner) [inline src/Grids/vertically_stretched_rectilinear_grid.jl:209]: 💡

**Review by glwagner — APPROVED** [review]

**Pull request merged**